core/src/upgrade/mod.rs

```diff
@@ -179,10 +179,10 @@ pub trait OutboundUpgradeExt<C>: OutboundUpgrade<C> {
 
     /// Returns a new object that combines `Self` and another upgrade to support both at the same
     /// time.
-    fn or_outbound<U>(self, upgrade: U) -> OrUpgrade<Self, U>
+    fn or_outbound<U, O, E>(self, upgrade: U) -> OrUpgrade<Self, U>
     where
         Self: Sized,
-        U: OutboundUpgrade<C, Output = Self::Output, Error = Self::Error>
+        U: OutboundUpgrade<C, Output = O, Error = E>
```

Comment:
I don't think we need to add the constraints here.

```suggestion
        U: OutboundUpgrade<C>
```

(also remove the newly-added generics)